### ISSUE TITLE
UX (sessions) : Improved session list

### ISSUE THREAD

**Issue by ciyer** [SwissDataScienceCenter/renku-ui#1219]
## Description

Make the session listing more powerful. We can drop the session list from the project page -- there should just be one global session list.

## Mock-ups

![image](https://user-images.githubusercontent.com/1196411/107025617-f86da000-67a9-11eb-8363-d443eff02597.png)

## API

The UI will need to able to get from the session the following information:

- uncommitted work yes/no?
- clone ahead/behind origin?
- last push to autosave branch

And the UI will need to be able to invoke the following commands on the session:
- push to autosave branch
- renku save (with message provided)

## Progress

- [x] API design (sprint-2021-02-19)
- [ ] Status handling
- [ ] Saving

**Comment by rokroskar** [comment]
Just an idea - but maybe clicking on "sessions" in the project could take the user to the global project page but restrict the search to show only the project they just came from? With an obvious way to remove the filter so they can see all the sessions.

**Comment by rokroskar** [comment]
regarding "last push to autosave branch" - this implies that these pushes are periodic, but this is not the case. Or did you have something else in mind?

**Comment by rokroskar** [comment]
I just tried accessing git information via curl using the session cookie and the git extension endpoint and it works  - so I don't think this issue is really blocked by https://github.com/SwissDataScienceCenter/renku-python/issues/1659

**Comment by ciyer** [comment]
- That's a good idea to use the git extension endpoint. 
- The pushes to the autosave branch should be periodic.

**Comment by rokroskar** [comment]
We decided to not make them periodic initially for a few reasons, but the main one is that in the case that large files need to be pushed it could have detrimental effects on the user's session. In order to do periodic pushes, you need to have a mirror of the repo somewhere because you need to change the branch, commit, and push. I don't see a good way of doing this automatically, but maybe I'm missing something obvious?

**Comment by ciyer** [comment]
If we do not intend to save periodically, then it should not be called "autosave". We should change it to "recovery" or something similar.

**Comment by rokroskar** [comment]
sure, we can give it a different name - the more important question I think is whether we want to enable periodic saves or not. Maybe we can actually do better by having dedicated re-usable volumes instead, for example.

**Comment by ciyer** [comment]
Note for implementor: check that the API provided by the Jupyter git extension in available if the user has launched RStudio (it probably is, but we should just test).

**Comment by rokroskar** [comment]
@ciyer it's available if the git jupyterlab extension is installed. Eventually we will probably want to change this to have a sidecar container running so that the access to git functionality is independent of what is in the user's docker image. But given that 99% of users use our base images I think this should be fine for now, provided the error handling is robust.

**Comment by lorenzo-cavazzi** [comment]
Here is the reference for the [GitLab extension endpoints](https://github.com/jupyterlab/jupyterlab-git/blob/master/jupyterlab_git/handlers.py#L787-L819) [[permalink here](https://github.com/jupyterlab/jupyterlab-git/blob/e1c42c6f970d1ba6b26a25a792397e444d121d81/jupyterlab_git/handlers.py#L787-L819) - just in case]

Having access to the current git status opens up new possibilities.
Instead of the enhanced autosave, we could implement a "commit and push" button, leaving to the autosave its current unsaved work recovery role.

If we decide to rely on those API, we should build support for them in renku notebooks.
Since we are currently polling with high frequency for the pod status (every 3s), we should first check if that causes potential problems and if the response is always accurate. If that works fine, we should include this information in the `GET /servers` endpoint.
As mentioned, the UI should handle the missing pieces correctly since we actually expect them to be missing in specific circumstances (e.g. when the pods are starting/terminating).

I'm not sure what we should show in `branch` and `commit`. I believe the best thing would be to show the current ones as provided by the GitLab extension API. Still, we should show a warning icon (with a short explanation in a popover) whenever the current commit is different from the one used to start the pods since it's possible the Dockerfile (or other files used in the pod startup) were different, therefore leading to potential differences in the environment compared to a brand new pod started from the same commit.

**Comment by lorenzo-cavazzi** [comment]
Should we temporarily put aside the `save` button and move on with the rest?
That part seems trickier and requires swissdatasciencecenter/renku-python#1659 to have the benefits of `renku save`. We could tackle it in a follow-up issue.

**Comment by lorenzo-cavazzi** [comment]
Tip: remember to check for PVS support using the new `GET /autosaves` endpoint and show a warning when stopping a session with unsaved work in a PVS enabled environment (the content will go lost) SwissDataScienceCenter/renku-notebooks/issues/584

**Comment by lorenzo-cavazzi** [comment]
The session pages have completely changed in the meanwhile, actually showing the information suggested in the initial post or some variation of it (E.G: pause instead of autosave)
